### ISSUE TITLE
Kafka client problems kill or hang clickhouse-server process or prevent restart

### ISSUE THREAD

**Issue by kmatt** [yandex/ClickHouse#5085]
After enabling SSL connections, some Kafka table clients are failing. This appears to demonstrate where those client failures cause problems in the parent clickhouse-server process.

**Describe the bug**

With Kafka table receiving initially messages as expected and successfully writing through materialized views, failures in the clients sometimes cause the server process to become unresponsive or terminate, and restarts of the clickhouse-server daemon sometimes fail.

The only messages that appear in the clickhouse-server.err.log are:

```
2019.04.22 23:50:02.805504 [ 1 ] {} <Error> Application: Caught exception while loading metadata: std::exception. Code: 1001, type: cppkafka::Exception, e.what() = Failed to create consumer handle: ssl.certificate.location failed: No error
2019.04.22 23:50:02.806111 [ 1 ] {} <Error> Application: Failed to create consumer handle: ssl.certificate.location failed: No error
2019.04.22 23:51:53.390405 [ 1 ] {} <Error> Application: Caught exception while loading metadata: std::exception. Code: 1001, type: cppkafka::Exception, e.what() = Failed to create consumer handle: ssl.certificate.location failed: No error
2019.04.22 23:51:53.390904 [ 1 ] {} <Error> Application: Failed to create consumer handle: ssl.certificate.location failed: No error
```

Only known way to start server is to remove metadata files for Kafka tables, start server and recreate consumer views.

**How to reproduce**
1) ClickHouse server version 19.1.14 on CentOS Linux release 7.4.1708 (Core)

```
CREATE TABLE kafka_1
(
    hash String,
    date UInt64,
    time UInt64,
    ...
)
ENGINE = Kafka
SETTINGS kafka_broker_list = 'msg1.*.com, msg2.*.com ,msg3.*.com, msg4.*.com', kafka_topic_list = 'logs', kafka_group_name = 'clickhouse0', kafka_format = 'JSONEachRow', kafka_row_delimiter = '\n', kafka_skip_broken_messages = 1, kafka_num_consumers = 1;
```

2) Verify messages can be recieved by `SELECT * LIMIT 1` from the above view.
3) Stop server with `service clickhouse-server stop`
4) Attempt to restart server with `service clickhouse-server start`
5) Note daemon is not running and see messages above emitted in logs

**Expected behavior**
Other nodes in this 4 node cluster start as expected and resume consuming messages

**Kafka configuration**

/etc/clickhouse-server/config.xml:

    <kafka>
        <max_poll_interval_ms>60000</max_poll_interval_ms>
        <session_timeout_ms>60000</session_timeout_ms>
        <heartbeat_interval_ms>10000</heartbeat_interval_ms>
        <reconnect_backoff_ms>5000</reconnect_backoff_ms>
        <reconnect_backoff_max_ms>60000</reconnect_backoff_max_ms>
        <security_protocol>SSL</security_protocol>
        <ssl_ca_location>/etc/clickhouse-server/ssl/kafka-ca-qa.crt</ssl_ca_location>
        <ssl_certificate_location>/etc/clickhouse-server/ssl/client_clickhouse_client.pem</ssl_certificate_location>
        <ssl_key_location>/etc/clickhouse-server/ssl/client_clickhouse_client.key</ssl_key_location>
        <ssl_key_password>pass</ssl_key_password>
    </kafka>

**Comment by kmatt** [comment]
It appears that SSL failures can bring a running server down as well. Just observed a node that was previously ingesting messages stop without after emitting only these two messages into the logs:

```
2019.04.23 00:08:49.359147 [ 1 ] {} <Error> Application: Caught exception while loading metadata: std::exception. Code: 1001, type: cppkafka::Exception, e.what() = Failed to create consumer handle: ssl.certificate.location failed: No error
2019.04.23 00:08:49.392926 [ 1 ] {} <Error> Application: Failed to create consumer handle: ssl.certificate.location failed: No error
```

**Comment by kmatt** [comment]
Switched to using Docker yandex/clickhouse-server:19.1.14, pointing to the same local data volumes and config.xml.

Upon starting containers, two started successfully, two emitted the following errors and terminated without any additional information:

```
2019.04.23 14:59:33.266127 [ 1 ] {} <Error> Application: Caught exception while loading metadata: std::exception. Code: 1001, type: cppkafka::Exception, e.what() = Failed to create consumer handle: ssl.certificate.location failed: No error
2019.04.23 14:59:33.290951 [ 1 ] {} <Error> Application: Failed to create consumer handle: ssl.certificate.location failed: No error

**Comment by kmatt** [comment]
Entire clickhouse-server.err.log from other failed node:

```
2019.04.23 14:59:32.878876 [ 2 ] {} <Error> BaseDaemon: ########################################
2019.04.23 14:59:32.879034 [ 2 ] {} <Error> BaseDaemon: (from thread 24) Received signal Segmentation fault (11).
2019.04.23 14:59:32.879061 [ 2 ] {} <Error> BaseDaemon: Address: 0xf336190
2019.04.23 14:59:32.879074 [ 2 ] {} <Error> BaseDaemon: Access: read.
2019.04.23 14:59:32.879087 [ 2 ] {} <Error> BaseDaemon: Address not mapped to object.
2019.04.23 14:59:33.124787 [ 2 ] {} <Error> BaseDaemon: 0. /usr/bin/clickhouse-server(arena_dalloc_bin_junked_locked+0x86) [0x6d902e6]
2019.04.23 14:59:33.124846 [ 2 ] {} <Error> BaseDaemon: 1. /usr/bin/clickhouse-server() [0x6d88294]
2019.04.23 14:59:33.124858 [ 2 ] {} <Error> BaseDaemon: 2. /usr/bin/clickhouse-server() [0x6d8992f]
2019.04.23 14:59:33.124870 [ 2 ] {} <Error> BaseDaemon: 3. /usr/bin/clickhouse-server(tsd_cleanup+0x49) [0x6d8c559]
2019.04.23 14:59:33.124883 [ 2 ] {} <Error> BaseDaemon: 4. /lib/x86_64-linux-gnu/libpthread.so.0(+0x6408) [0x7fe1a7680408]
2019.04.23 14:59:33.124896 [ 2 ] {} <Error> BaseDaemon: 5. /lib/x86_64-linux-gnu/libpthread.so.0(+0x781b) [0x7fe1a768181b]
```

**Comment by kmatt** [comment]
```
cat /var/log/clickhouse-server.err.log

2019.04.23 16:20:24.528645 [ 1 ] {} <Error> Application: Caught exception while loading metadata: std::exception. Code: 1001, type: cppkafka::Exception, e.what() = Failed to create consumer handle: ssl.ca.location failed: ../contrib/ssl/crypto/x509/by_file.c:254: error:0BFFF009:x509 certificate routines:CRYPTO_internal:PEM lib: , version = 19.5.3.8 (official build)
2019.04.23 16:20:26.364963 [ 1 ] {} <Error> Application: Failed to create consumer handle: ssl.ca.location failed: ../contrib/ssl/crypto/x509/by_file.c:254: error:0BFFF009:x509 certificate routines:CRYPTO_internal:PEM lib:
```

```
docker ps | grep clickhouse
1962a262e316        yandex/clickhouse-server:19.5.3.8                                                                 "/entrypoint.sh"         47 minutes ago      Up 17 minutes                           clickhouse19538
```

```
$ clickhouse-client
ClickHouse client version 19.1.14.
Connecting to localhost:9000.
Code: 210. DB::NetException: Connection refused (localhost:9000, 127.0.0.1)
```

**Comment by kmatt** [comment]
Similar occurring in v19.5.3.8 via Yandex Docker container:

```
2019.04.29 18:27:06.754492 [ 1 ] {} <Error> Application: Caught exception while loading metadata: std::exception. Code: 1001, type: cppkafka::Exception, e.what() = Failed to create consumer handle: ssl.ca.location failed: ../contrib/ssl/crypto/x509/by_file.c:254: error:0BFFF009:x509 certificate routines:CRYPTO_internal:PEM lib: , version = 19.5.3.8 (official build)
```

**Comment by kmatt** [comment]
@abyss7 - Similar occurring in v19.5.3.8 via Yandex Docker container:

```
2019.04.29 18:27:06.754492 [ 1 ] {} <Error> Application: Caught exception while loading metadata: std::exception. Code: 1001, type: cppkafka::Exception, e.what() = Failed to create consumer handle: ssl.ca.location failed: ../contrib/ssl/crypto/x509/by_file.c:254: error:0BFFF009:x509 certificate routines:CRYPTO_internal:PEM lib: , version = 19.5.3.8 (official build)
```

That was generated on restarting the Docker container. The container stopped ingesting from Kafka 12 hours prior, but there are no events at all logged in /var/log/clickhouse-server.err.log or in the Docker logs.

Anywhere else I can look for diagnostic information, or settings to increase log levels?

**Comment by kmatt** [comment]
So I could turn up logging to DEBUG level, I stopped all four Docker containers, and changed the logging level with the new log path. Upon restart of the containers, none of the 4 servers would accept clickhouse-client connections, and Kafka ingest rate was 0.

Only way to get them started was again to drop kafka table engines and consumer materialized views in the metadata path, which allowed servers to start normally. Re-creating Kafka consumers exactly as before resumed ingest.

**Comment by abyss7** [comment]
The problem is solved in v19.8+ - the current solution only allows to start server with broken Kafka tables (they won't retry to connect to Kafka brokers again) and re-create/remove Kafka table if required.
